### PR TITLE
Typography improvements

### PR DESCRIPTION
**Pull request by Marat-Tanalin**
Proper em dashes and quotes.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,26 +1,26 @@
 ## Contributing to the C++ Core Guidelines
 
->"Within C++ is a smaller, simpler, safer language struggling to get out." 
->-- <cite>Bjarne Stroustrup</cite>
+>“Within C++ is a smaller, simpler, safer language struggling to get out.”<br />
```
Comment (jwakely) [inline CONTRIBUTING.md]: You should be able to add spaces to the end of the line to add a line-break using markdown, not html

**Comment by jwakely** [comment]
N.B. there's no need to create a second pull request, if you push something different to your master branch your original request would have been updated.

File: CONTRIBUTING.md (hunk 1 above)
Comment (Marat-Tanalin) [inline CONTRIBUTING.md]: @jwakely Originally I tried to use `\` at the end of line (it should create a line break in Markdown), but it didn’t work on GitHub for some reason. Could you provide a working example of adding a line break to an *.md file with no HTML?

P.S. Looks like different Markdown parsers are used for GitHub comments and *.md files:
in comments, line break is added automatically just based on new-line character while in *.md files new line character is ignored.

**Comment by Marat-Tanalin** [comment]
@jwakely
> there's no need to create a second pull request, if you push something different to your master branch your original request would have been updated.

Oh, thanks. Didn’t know about that.

**Comment by blakehawkins** [comment]
It should be the renderer's job to format these. Suggest close.

**Comment by Marat-Tanalin** [comment]
@blakehawkins Automatic formatting is error-prone and only somewhat reasonable when verified manually by a human anyway.

**Comment by blakehawkins** [comment]
See https://travis-ci.org/isocpp/CppCoreGuidelines/builds/207216964#L1132

It's much easier to improve a renderer than it is to require that changes to the doc use proper dashes and quote characters

**Comment by Marat-Tanalin** [comment]
@blakehawkins
> See https://travis-ci.org/isocpp/CppCoreGuidelines/builds/207216964#L1132

How should I treat this? I suspect this is an automated check intended to check _code_ while C++ Core Guidelines are a book-like document, not code. Fwiw, please note that I didn’t touch the _code_ examples (both block-level and inline) in any way, only text parts are formatted of course.

> It's much easier to improve a renderer than it is to require that changes to the doc use proper dashes and quote characters

It’s impossible to create a 100% non-error-prone automatic formatter (even aside from that such formatting is currently _not_ applied to C++ Core Guidelines anyway, and is unlikely going to in the foreseeable future).

An automation tool could _help_ in generating some _draft_ formatting that a human could then verify, but it cannot be the _only_ thing that we could rely on (and I already formatted the entire document manually). I’m in web development since 2002, quality text formatting is an important part of a quality website, so I know quite well what I am talking about.

**Comment by blakehawkins** [comment]
> How should I treat this? I suspect this is an automated check intended to check code while C++ Core Guidelines are a book-like document, not code. Fwiw, please note that I didn’t touch the code examples (both block-level and inline) in any way, only text parts are formatted of course.

That isn't what it is. Remark is for linting markdown quality. See https://github.com/wooorm/remark-lint

> It’s impossible to create a 100% non-error-prone automatic formatter (even aside from that such formatting is currently not applied to C++ Core Guidelines anyway, and is unlikely going to in the foreseeable future).

That's a pretty bold statement. Usually if you have a renderer that uses a real parser and it can't display something correctly, it points to a problem in the input data.

I'm not sure what the second clause in your sentence is trying to say.

> An automation tool could help in generating some draft formatting that a human could (and should) then verify manually, but it couldn’t be the only thing we could rely on. I’m in web development since 2002, quality text formatting is an important part of a quality website, so I know quite well what I am talking about.

Markdown isn't a "webpage format" and we aren't doing "web development". Real books are rendered using text rendering engines.

Take a look at what `pandoc` + `xelatex` make of the most recent successful build (`8a9bb3bc61a97aaf9a649e07dcdeaa28daf20407`):

https://www.docdroid.net/z4FvVsi/out.pdf.html#page=25

> You are free to create a better renderer in the future (though I’m not sure how this could influence the Markdown renderer built into GitHub), but the entire current text is already carefully formatted by me manually, and there is no reason to drop this work.

I realize it's painful to drop hard work, but it seems that you made a thousand changes to this doc without understanding what the CI build does. Is that fair?

**Comment by Marat-Tanalin** [comment]
@blakehawkins

> Remark is for linting markdown quality.

If a tool recommends to use wrong keyboard characters in UTF-8 _text_ (not code) instead of correct typographic ones, the tool is just plain wrong and harmful. There is no such thing as “markdown quality” outside of its _syntax_ validity. Checking text for being ANSI-only (we are in 2017 with UTF-8 everywhere, hello?) is out of MD-linter responsibility, so its recommendations as for text (not code) are in no way something that should be taken into account.

> it points to a problem in the input data

This is it. Problem in the input data is a broad term. You can never guarantee the input data is perfectly correct from a specific automated-tool perspective. For example, if an inline code fragment like `i--` wasn’t marked-down as code by wrapping it in back-ticks, the renderer could blindly replace `--` with `—`, thus making the text incorrect and confusing: `i—`.

> Markdown isn't a "webpage format" and we aren't doing "web development". Real books are rendered using text rendering engines.

That looks like an offtopic, but afaik, real books aren’t usually written (at least commercially) in Markdown. Markdown is certainly not a webpage format (HTML is), it’s a _content_ format that just somewhat more low-level and garbage-free than WYSIWYG while still being simple enough to use by nonadvanced authors compared with HTML.

> it's painful to drop hard work

It’s not just hard work, it’s work _nobody_ will probably _ever_ do for you again anymore. Given that the guidelines are now apparently mostly complete, its current typography state (after accepting my pull request) would cover the most of the text even if subsequent commits still didn’t account for typography (moreover, I could just make minor corrections periodically to maintain high typography-wise quality of the Guidelines).

> understanding what the CI build does

I’m aware of what continuous integration is. And I believe the actual reason why you are against the change is _exactly_ that it conflicts with other pending pull requests. I see the two solutions for preventing such conflicts:

* making such massive changes only when there is no previous unmerged pull requests at all. Almost impossible — there are almost always some pull requests;

* VCS should be able to diff on _per-character_ (instead of per-line) basis. This either should be already enableable via Git settings (and available at least for users of paid GitHub accounts) or is a good way for Git to be improved. This way, VCS could easily merge most of pull requests with no conflict issues at all.

**Comment by blakehawkins** [comment]
Hi again,

> If a tool recommends to use wrong keyboard characters in UTF-8 text (not code) instead of correct typographic ones, the tool is just plain wrong and harmful.

So, that isn't quite right. The tool's recommendations are very much dependent on the config we provide - see https://github.com/isocpp/CppCoreGuidelines/blob/master/scripts/nodejs/remark/.remarkrc

I think you'll find that your *opinion* on use of proper UTF-8 characters will be a bit contentious. I'm sure you can think of more than one argument for why the use of straight quotes and hyphens only *might* be preferred by someone, no?

> There is no such thing as “markdown quality” outside of its syntax validity.

That isn't right either. Markdown quality includes things like consistency (are all your code blocks fenced the same way?) and flavor preference (github flavored markdown in this case).

> Checking text for being ANSI-only (we are in 2017 with UTF-8 everywhere, hello?) is out of MD-linter responsibility, so its recommendations as for text (not code) are in no way something that should be taken into account.

Objection: appeal to novelty.

The value in using ANSI-only text in a text data format is that this document is maintained by people typing with keyboards. If your keyboard has proper quote characters and em-dashes, that's cool, but it isn't standard. Don't forget that your target audience here is quite a lot of people who still write (everything) in vim. Copy-pasting utf characters is not something people are going to like.

If you really do think proper UTF characters are right here, probably the best course of action would be to push for changes to the remark config, so that proper UTF characters are *consistently* preferred, and therefore straight quotes and double hyphens are flagged by the CI build.

I don't think people will go for that, though.

> This is it. Problem in the input data is a broad term. You can never guarantee the input data is perfectly correct from a specific automated-tool perspective. For example, if an inline code fragment like i-- wasn’t marked-down as code by wrapping it in back-ticks, the renderer could blindly replace -- with —, thus making the text incorrect and confusing: i—.

Do you really mean to say that we can push for UTF quotes and em-dashes, but can't push to ensure all code (inline or block) is wrapped in markdown code formatting? I think that's an absurd claim.

> Also, as I said, you could most likely not influence the existing MD renderer of GitHub itself in any way even if a third-party renderer could be able to produce correct typography-wise result.

I'm not really interested in changing the github renderer.

> That looks like an offtopic, but afaik, real books aren’t usually written (at least commercially) in Markdown. 

I don't think this is really a valid point. LaTeX will also flag "correct" quote characters and em-dashes as invalid.

> It’s not just hard work, it’s work nobody will probably ever do for you again anymore.

Surely the computer will do it for me again when I write `pandoc -r markdown -w latex --latex-engine=xelatex CppCoreGuidelines.md` - just the same way as it did in the link above.

> I’m aware of what continuous integration is.

So, sorry if that wasn't clear. I didn't mean to insinuate that you don't know what CI is (actually, I wrote "the CI build", not "CI builds". My point is that you don't seem to know the expectations for what this particular CI build checks.

> And I believe the actual reason why you are against the change is exactly that it conflicts with other pending pull requests. I see the two solutions for preventing such conflicts:

That is not correct. The CI build, if anything, would compare your build against the master branch.

It is definitely true that the current master branch is failing the CI build, though. My objection is not that your PR is failing (it's not your job to necessarily fix existing remark moans) - my objection is that your changes intentionally make existing passing text *stop passing*.

> making such massive changes only when there are no previous unmerged pull requests at all. Almost impossible — there are almost always some pull requests;

That wouldn't be necessary at all. Your branch doesn't have any conflicts, just new CI moans.

I hope this clears some of your confusion.

--

Finally, two suggestions from myself:

* to prevent this from happening in future, it would really have been good to open an issue first to suggest the changes, before actually performing them. That's what's suggested here - https://github.com/isocpp/CppCoreGuidelines/blob/master/CONTRIBUTING.md#pull-requests

* It might be worth making a new branch (perhaps called `proper_utf_chars`?) with *different remark settings* so you can consistently maintain proper quote characters and em dashes. I bet the maintainers of this repo would be happy to host that branch here, officially, so long as you were willing to maintain it.

**Comment by Marat-Tanalin** [comment]
@blakehawkins

> Objection: appeal to novelty.

Unicode has nothing to do with novelty. Unicode is our objective reality for at least 15+ years (e.&thinsp;g. Windows XP released in 2001 already had Unicode support).

Sorry, preferring ANSI over Unicode today looks like over-old-school’ish divorcement from reality I can’t help with. But that’s your right and choice.

> If your keyboard has proper quote characters and em-dashes, that's cool, but it isn't standard.

Fwiw, I have a regular keyboard like you or others have. I typically insert special characters via `Alt + 4-digit code` key sequences (e.&thinsp;g. `Alt+0151` for em dash) available at least in Windows.
> Objection: appeal to novelty.

Unicode has nothing to do with novelty. Unicode is our objective reality for many (probably at least 15 or so) years.

> If your keyboard has proper quote characters and em-dashes, that's cool, but it isn't standard.

Fwiw, I have an absolutely regular keyboard like you or others have. I typically insert special characters via `Alt + 4-digit code` key sequences (e.&thinsp;g. `Alt+0151` for em dash) available at least in Windows.

> Do you really mean to say that we can push for UTF quotes and em-dashes, but can't push to ensure all code (inline or block) is wrapped in markdown code formatting?

Like with the example I’ve provided, that’s called human factor. Anything is possible when input data are provided and reviewed by a human.

> I'm not really interested in changing the github renderer.

But the truth is that people most likely tend to read exactly the _GitHub_ version of Guidelines that in fact has poor typography.

> Surely the computer will do it for me again when I write `pandoc -r markdown -w latex --latex-engine=xelatex CppCoreGuidelines.md` - just the same way as it did in the link above.

Since this command requires a complex toolchain besides pandoc itself, I can’t say for sure what the result it should have.

Could you clarify whether the command you’ve provided does automatically replace ANSI surrogates for characters like quotes and em dashes to corresponding correct typographic characters (the command itself seems not to have an explicit option to enable such replacement, so it’s unclear)?

> I bet the maintainers of this repo would be happy to host that branch here, officially, so long as you were willing to maintain it.

No, sorry, I don’t have time resources and motivation for that.

**Comment by MikeGitb** [comment]
I don't care much one way or the other but out of curiosity: What exactly is the reason behind the "use straight quotes instead" rule?

@Marat-Tanalin: blakehawkins provided a link to the pdf created by the command: https://www.docdroid.net/z4FvVsi/out.pdf.html#page=25

**Comment by Marat-Tanalin** [comment]
@MikeGitb It’s unclear whether the PDF is generated using _default_ settings or with some explicit typography-related flags (I don’t see such flags in the command provided by @blakehawkins, hence my question).

**Comment by blakehawkins** [comment]
@Marat-Tanalin 

> Unicode has nothing to do with novelty. Unicode is our objective reality for at least 15+ years (e. g. Windows XP released in 2001 already had Unicode support).

> we are in 2017 with UTF-8 everywhere, hello?

*Appeal to Novelty* doesn't mean that I think unicode is a novelty - it means that I object to your argument on the grounds that it's a logical fallacy to imply that something newer is necessarily better in all contexts. https://en.wikipedia.org/wiki/Appeal_to_novelty

> Like with the example I’ve provided, that’s called human factor. Anything is possible when input data are provided by a human.

I think the whole point is that changes to this document are reviewed by other humans. I think you'd have a hard time finding code examples in this doc that aren't wrapped in the proper code blocks.

> Since this command requires a complex toolchain besides pandoc itself, I can’t say for sure what the result it should have.

The result should be consistent, and this is just one example of a tool for rendering markdown.

> Could you clarify whether the command you’ve provided does automatically replace ANSI surrogates for characters like quotes and em dashes to corresponding correct typographic characters (the command itself seems not to have an explicit option to enable such replacement, so it’s unclear)?

The command is just invoking `LaTeX` (or in this case, xelatex) and wrapping the markdown doc in a latex template.

LaTeX expects all quote characters as straight quotes, and typesets them in publish quality. Of course, as you mentioned, for printing and distributing such a thing, a large company like O'Reilly would probably have a human editor verify the quality as well.

One final note on this: as mentioned above, I used a particular commit to generate that pdf (where the build was *passing*). If I did it with your PR, I suspect it would probably fail.

> No, sorry, I don’t have time resources and motivation for that. Aside from that I consider maintaining a separate typography-correct version totally unreasonable: the main version itself should be as correct and readable as possible.

I think this is a key point. You don't have time to maintain consistency in this style, and everyone else has *neither* the time *nor* the interest to maintain it.

> I don't care much one way or the other but out of curiosity: What exactly is the reason behind the "use straight quotes instead" rule?

I think @tkruse might have a better answer, but I suspect the rule is for consistency. It's much more desirable to have 100% consistent straight quotes, than to have some straight quotes and some curly ones. Straight quote are also less likely to rot since someone modifying the document doesn't have to be interested/capable of maintaining the curly ones. Finally, straight quotes are the standard charset expected when typesetting with a tex engine.

> It’s unclear whether the PDF is generated using default settings or with some explicit typography-related flags

Just to repeat for clarity: there are no special typography flags; `tex` expects straight quotes.

**Comment by tkruse** [comment]
I added the rule for straight quotes after finding that some 1% of quotes were non-straight, and  inconsistency is worse than anything. But also in Markdown, the straight quotes in non-preformated text are the correct representation (coding) of typographic quotes in the rendering results, I think.

Github chose to use rendering of quotes and dashes in Markdown without transformation as the default for __viewing files in the source tree__ presumably because that view is targeted at contributors to the document, who commonly need to the the raw source of documents. Not for readers. That's not a github feature, not a bug.

For github, the correct channel for actual readers of the document are github pages or any other form of presentation that went through a publishing step. I helped this project to set up a gh-pages branch for that effect, it is linked at the top of the github pages (maybe not prominently enough).
The jekyll renderer does the transformation of quotes and dashes correctly, I think, check: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines
(Note that inside code blocks, double-minus and quotes remain as in the source)

So I believe this PR provides a technical solution where there is no technical problem.

Possibly the rendered github page could be advertized more. As an example, the github-pages version is not mentioned in the ```README.md```, and not visible on the page at all when opening ```CppCoreGuidelines.cpp```. Also I think @Marat-Tanalin could probably make that one look much better than I did.

But the ```CppCoreGuidelines.cpp``` files should probably remain as it is now with respect to quotes and dashes.

**Comment by jwakely** [comment]
Personally I don't think this is worth doing. I don't want to have to avoid using simple quote characters when editing the document.

**Comment by Marat-Tanalin** [comment]
@blakehawkins
> Appeal to Novelty doesn't mean that I think unicode is a novelty - it means that I object to your argument on the grounds that it's a logical fallacy to imply that something newer is necessarily better in all contexts.

Unicode is better _not_ because it’s newer, but because it is _universal_, extremely widely supported and is our objective reality. Again, it has _nothing to do_ with appeal to novelty.

> I used a particular commit to generate that pdf (where the build was passing). If I did it with your PR, I suspect it would probably fail.

I have no idea why it should fail, other than just due to a current automated-check rule (that could easily be removed) that forces quotes to be necessarily straight.

> I think this is a key point. You don't have time to maintain consistency in this style, and everyone else has neither the time nor the interest to maintain it.

Maintaining a _separate_ version existing _besides_ the main _permanently-unformatted_ version would apparently involve making the _same_ typography corrections _over and over again_ (be it manually or automatically) — _that_’s what I consider unreasonable.

On the contrary, once the corrections I’ve _already_ made are merged with the _main_ version of the document, I could easily make additional _minor_ corrections periodically, while the document would already always be 99% typography-correct (and 100% _up-to-date_ at the same time unlike a possible entirely separate version) even with no any typography-wise maintenance at all.

> Straight quote are also less likely to rot since someone modifying the document doesn't have to be interested/capable of maintaining the curly ones.

I have no idea how the proper typography already existing in the document could impact contributors in a bad way regardless of whether they are interested/capable of maintaining anything.

> straight quotes are the standard charset expected when typesetting with a tex engine.

As long as a tex engine supports Unicode (and it obviously can’t not to), that probably doesn’t matter.

@tkruse
> I added the rule for straight quotes after finding that some 1% of quotes were non-straight, and inconsistency is worse than anything.

Well, _inconsistently good_ (that the document would be once my pull request is merged) is probably much better than _consistently bad_.

> Github chose to use rendering of quotes and dashes in Markdown without transformation as the default for viewing files in the source tree presumably because that view is targeted at contributors to the document, who commonly need to the the raw source of documents. Not for readers.

Fwiw, exactly as a _reader_, I prefer to use the GitHub version just to get the most recent version not outdated in any way.

> Note that inside code blocks, double-minus and quotes remain as in the source

That’s quite expected given that corresponding code parts are properly marked-down. As I said, my example was solely about possible code _not marked-down explicitly_.

> I think @Marat-Tanalin could probably make that one look much better than I did.

Could you clarify how exactly this could be done? Thanks.

**Comment by tkruse** [comment]
@Marat-Tanalin : Unless you find (many) other people to support your view, I think you should get used to the idea that the PR wont get merged.

There are alternatives to achieve what you want. One is to use the master branch to render the github-pages, that way the latter is always up-to-date. Another is to use CI for publishing, that could publish both HTML and PDF as an example, for every new version. Lastly it is possible to write an html5 app that can read any raw version of the document and render it to  pretty html on the fly. None is easy to achieve, but all of them are far more likely to happen then your PR being merged, I think.
And there is the alternative of you reading the file on your local machine rendering it locally in whatever way you like, updating to the latest version is then as cheap as git pull.

I personally prefer the first alternative, and since github added features to manage the rendering from a docs folder on the master branch, there is a slim chance the core maintainers may accept a PR with that effect:
https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/

If you want to help improving the version published at http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines , you can check out the ```gh-pages``` branch, and then edit the HTML templates and CSS.

There are github tutorials on how to use jekyll locally to render locally, but you can also just create a gh-pages branch in your repository, and github will create a rendered site at http://Marat-Tanalin.github.io/CppCoreGuidelines/CppCoreGuidelines whenever you push that branch to your repository.
When making a PR, you then would have to take care to make it against the 'gh-pages' branch of isocpp/CppCoreGuidelines, as opposed to 'master', which is the default.

**Comment by AndrewPardoe** [comment]
I'm favoring the simpler formatting, based on recommendations of @tkruse, @blakehawkins, @jwakely. People edit this document with a number of editors--I'd prefer to keep the text as simple as possible. Unfortunately, we live in a world where ASCII rules.

**Comment by Marat-Tanalin** [comment]
@AndrewPardoe
That’s OK. Thanks to all participated in the thread.

I would like to summarize the points I stated in the comments above:

* Unicode is today’s objective reality, _not a novelty_ in any way. Unicode is better _not_ because it’s newer, but because it is _universal_ and makes documents looking better and more readable with no need for extra pre-/post-processing while being extremely widely supported.

* No special keyboard is needed for inserting non-keyboard characters. For example, in Windows, such characters can be inserted via `Alt + 4-digit code` key sequences (e.&thinsp;g. `Alt+0151` for em dash).

* Converting Markdown to PDF requires a _complex toolchain_ not everyone can properly set up and use.

* _99% good_ is much better than _100% bad_. Once the corrections I’ve made are merged with the main version of the document, the document would always be 99% typography-correct (and 100% up-to-date at the same time unlike a possible entirely separate version) even with no any typography-wise maintenance at all.

> we live in a world where ASCII rules.

Using ASCII should be reasonable. It sometimes is in _code_, but typically not in regular _text_. I’m curious what you would do if the document was written in a _nonlatin language_ like Russian/Cyrillic.

@tkruse
> There are alternatives to achieve what you want. One is to use the master branch to render the github-pages, that way the latter is always up-to-date.
> I personally prefer the first alternative

Fwiw, I like this option too. What could we do to make that happen?

**Comment by tkruse** [comment]
@Marat-Tanalin : The first step would be to validate that my idea works. Github allows using the master branch for docs, and allows using a doc folder to keep the ugly stuff aside a bit. 

But I now remember that the github-pages branch always needs an additional conversion step, e.g. see:
https://github.com/isocpp/CppCoreGuidelines/blob/gh-pages/INSTRUCTIONS.txt
Step 5. and 6.

I cannot vouch that there is a solution.

This can be done on any fork of this repo (any github account).

Once there is a viable solution, this has to be proposed as a ticket. I am not sure whether the core maintainers would accept a "doc" folder in the master branch of the repository root that exists only for github rendering.

**Pull request closed without merging**